### PR TITLE
docs: add information about lifespan context

### LINKED ISSUE #2403: Docs: order of lifespan ctx exit and on_shutdown hooks
### Summary

From a quick glance at the docs, it doesn't appear that we explicitly document the order that `__aexit__` methods on the `lifespan` context managers are called, relative to the registered `on_shutdown` hooks. This ordering might be important, especially that it prevents an `on_shutdown` hook from interacting with a resource that is managed by the lifespan context, e.g., a db engine, or redis connection or similar.

Take this app, for ex:

```py
app = Litestar(lifespan=[ctx_a, ctx_b], on_shutdown=[hook_a, hook_b])
```

On shutdown, the order of calls is:
1. `ctx_b.__aexit__()`
2. `ctx_a.__aexit__()`
3. `hook_a`
4. `hook_b`

So, on shutdown, context managers are exited in reverse order (which is natural behavior of the exit stack), and before shutdown hooks. Then, shutdown hooks are called in registered order.

We should also add tests to lock this behavior in, as FWICT (checked here: https://github.com/litestar-org/litestar/blob/main/tests/unit/test_asgi_router.py), we have tests that they are all called on shutdown, but nothing that tests for explicit order.

---

## Funding
* If you would like to see an issue prioritized, make a pledge towards it!
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/litestar-org/litestar/issues/2403">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/2403/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by orgarten**
### Pull Request Checklist

- [ ] New code has 100% test coverage
- [ ] (If applicable) The prose documentation has been updated to reflect the changes introduced by this PR
- [ ] (If applicable) The reference documentation has been updated to reflect the changes introduced by this PR
- [x] Pre-Commit Checks were ran and passed
- [x] Tests were ran and passed

### Description

- Added information regarding the order of execution if multiple lifespan context managers and shutdown hook are specified
- Added tests to ensure the order

### Close Issue(s)
Closes #2403

### CODE REVIEW

File: docs/usage/applications.rst [hunk 1]
```diff
@@ -65,6 +65,28 @@ connection, around.
     :language: python
     :caption: Handling a database connection
 
+Order of execution
+-----------------------------------------------------
+
+When multiple lifespan context managers and ``on_shutdown`` hooks are specified Litestar will invoke the context
+managers in inverse order before the shutdown hooks are invoked.
+
```
Comment (JacobCoffee) [inline docs/usage/applications.rst:73]:
```suggestion

When multiple lifespan context managers and ``on_shutdown`` hooks are specified, Litestar will invoke the context
managers in inverse order before the shutdown hooks are invoked.

```

File: docs/usage/applications.rst [hunk 2]
```diff
@@ -65,6 +65,28 @@ connection, around.
...
+Let's say there are two lifespan context managers ``ctx_a`` and ``ctx_b`` as well as two shutdown hooks ``hook_a`` and
```
Comment (JacobCoffee) [inline docs/usage/applications.rst:74]:
```suggestion
Consider the case where there are two lifespan context managers ``ctx_a`` and ``ctx_b`` as well as two shutdown hooks ``hook_a`` and
```
I think this is easier for ESL people

**Review by JacobCoffee — APPROVED** [review]

File: docs/usage/applications.rst (hunk 2 above)
Comment (orgarten) [reply docs/usage/applications.rst]: Great point. I am ESL, too, so I am never too sure about how it sounds to native speakers or how digestible it is for others ;)

**Comment by JacobCoffee** [comment]
Sorry i didnt realize it was draft before I reviewed :P

**Comment by orgarten** [comment]
> Sorry i didnt realize it was draft before I reviewed :P

It's all good ;) Helpful feedback anyway :) The tests for this should go into `tests/unit/test_app.py` or is there a better place? (see my latest comment in #2403)

**Comment by orgarten** [comment]
@all-contributors please add @orgarten for code

**Comment by orgarten** [comment]
@all-contributors please add @orgarten for docs

**Comment by JacobCoffee** [comment]
@all-contributors please add @orgarten for test

**Comment by JacobCoffee** [comment]
I do not think that the pyright failure is related to you, mering

**Comment by orgarten** [comment]
Thank you! The pyright failure is fixed by #2605

**Pull request merged**